cpp/parquetjni/parquetjni.cc

```diff
@@ -229,46 +262,85 @@ class DatasetBatchReader : public arrow::RecordBatchReader {
     // Access private fields
     DatasetBatchReader *reader = static_cast<DatasetBatchReader *>(out->get());
     // Load the first file and get the schema.
-    ARROW_RETURN_NOT_OK(reader->LoadNext());
-    reader->schema_ = reader->current_reader_->schema();
+    ARROW_RETURN_NOT_OK(reader->EnqueueReads(1));
+    if (reader->file_readers_.empty()) {
+      return arrow::Status::Invalid("No files in dataset");
+    }
+    TRACE("ParquetJniDataset::Open::LoadNext", profile_start, load_end);
+    std::shared_ptr<arrow::RecordBatchReader> batch_reader;
+    RETURN_NOT_OK(reader->file_readers_.front().get().GetReader(&batch_reader));
+    reader->schema_ = batch_reader->schema();
+    TRACE("ParquetJniDataset::Open::Schema", load_end, open_end);
     return arrow::Status::OK();
   }
 
   arrow::Status LoadNext() {
-    current_reader_ = nullptr;
-    if (current_jni_reader_) {
-      RETURN_NOT_OK(current_jni_reader_->Close());
+    if (!file_readers_.empty()) {
+      VLOG(2) << "Closing a file";
+      BufferedFile &buffered_file =
+          const_cast<BufferedFile &>(file_readers_.front().get());
```

Comment:
is there a pop_front()?

Reply:
Unfortunately no. https://en.cppreference.com/w/cpp/container/queue

Reply:
ok